src/ngraph/ngraph_stats.cc

```diff
@@ -0,0 +1,146 @@
+/*******************************************************************************
+* Copyright 2018 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "ngraph_stats.h"
+
+#include <iomanip>
+#include "ngraph_utils.h"
+
+namespace ngraph_bridge {
+
+std::string exe_mode_to_string(int mode) {
+  switch (mode) {
+    case 0:
+      return "Inference";
+    case 1:
+      return "Train";
+    default:
+      return std::to_string(mode);
+  }
+}
+void NGraphStats::dump(std::ostream& out) {
+  if (ngraph_log_timer()) {
+    // accumulator for forward/backward/Combined summary at the end
+    const int pass_count = 3;
+    const std::string pass_name[pass_count] = {"Forward", "Backward",
+                                               "Combined"};
+    std::vector<ngraph::runtime::PerformanceCounter> pass_perf[pass_count];
+
+    // iterate all the graphs and print their performance stats
+    for (auto& g : graphs_) {
+      if (g != nullptr) {
+        out << std::string(total_margin_, '#') << "\n";
+        out << "# Graph " << g->name_ << std::endl;
+        auto backend = GetBackendFromContext(g->context_);
+
+        auto print_perf_for_pass = [&](
+            const std::shared_ptr<ngraph::Function>& func, const int pass) {
+          std::vector<ngraph::runtime::PerformanceCounter> perf_data =
+              backend->get_performance_data(func);
+          if (perf_data.size() > 0) {
+            out << std::string(total_margin_, '-') << "\n";
+            out << "# " + pass_name[pass] << std::endl;
+            print_perf_data(out, perf_data);
+            pass_perf[pass].insert(pass_perf[pass].end(), perf_data.begin(),
+                                   perf_data.end());
+          }
+        };
+
+        // output inference/training execution mode
+        for (int i = 0; i < kGraphExeModeCount; ++i) {
+          out << std::string(total_margin_, '=') << "\n";
+          out << "# Mode: " << exe_mode_to_string(i) << std::endl;
+          print_perf_for_pass(g->ngraph_forward[i], 0);
+          print_perf_for_pass(g->ngraph_backward[i], 1);
+        }
+      }
+    }
+
+    auto print_pass_summary = [&](const int i) {
```

Comment:
only used once, why is this a lambda, just curious.

Reply:
Right, that's the result of a few refactors. It could be just inside the for loop below.